### PR TITLE
.so symlink, artifact, setup changes

### PR DESCRIPTION
**Pull request by greyblue9**
- Setup.py
    + retrieve version using 'ast' instead of directly importing, for the case where the dependencies are not yet installed
        e.g.    python3 setup.py develop
    + add 'requires' dependencies to compliment 'install_requires'
- src/use/use.py
    + fixes for solib handling
    + artifact matcher functions simplified

### CODE REVIEW

File: src/use/use.py [hunk 1]
```diff
@@ -177,11 +180,11 @@ def fail_or_default(default, exception, msg):
     else:
         raise exception(msg)
 
-def apply_aspect(mod:ModuleType, check:callable, pattern:str, decorator:callable):
+def apply_aspect(mod:ModuleType, check:callable, pattern:str, decorator:Callable[...]):
     """Apply the aspect as a side-effect, no copy is created."""
     # TODO: recursion?
     parent = mod
-    for name, obj in parent.__dict__.items():
+    for _, obj in parent.__dict__.items():
```
Comment (amogorkon) [inline src/use/use.py:187]: why not .values() if we're on it?

File: setup.py [hunk 2]
```diff
@@ -11,14 +9,19 @@
 src = os.path.join(here, "src/use")
 sys.path.insert(0, src)
 
-import use
+import ast
+with open(os.path.join(src, "use.py")) as f:
+    mod = ast.parse(f.read())
+    version = [*filter(lambda n: isinstance(n, ast.Assign) \
```
Comment (amogorkon) [inline setup.py:15]: Comment please :)

File: src/use/use.py [hunk 3]
```diff
@@ -59,7 +59,6 @@
 import atexit
 import codecs
 import configparser
-import functools
```
Comment (amogorkon) [inline src/use/use.py:62]: we'll need functools again, at least for wraps i guess

File: src/use/use.py [hunk 4]
```diff
@@ -177,11 +180,11 @@ def fail_or_default(default, exception, msg):
     else:
         raise exception(msg)
 
-def apply_aspect(mod:ModuleType, check:callable, pattern:str, decorator:callable):
+def apply_aspect(mod:ModuleType, check:callable, pattern:str, decorator:Callable[...]):
```
Comment (amogorkon) [inline src/use/use.py:183]: nice catch, thanks

File: src/use/use.py [hunk 5]
```diff
@@ -378,32 +382,27 @@ def is_version_satisfied(self, info:Union[dict,str]):
         sv = Version(".".join(map(str, sys.version_info[0:3])))
         vstr = info if isinstance(info,str) \
                   else info["requires_python"] \
-                    or info["python_version"]
+                    or "==%s" % info["python_version"]
```
Comment (amogorkon) [inline src/use/use.py:385]: f-str might be nicer

**Review by amogorkon — CHANGES_REQUESTED** [review]

File: src/use/use.py (hunk 1 above)
Comment (greyblue9) [reply src/use/use.py:187]: @amogorkon Good question, I dont have any memory of making this change! But I like your suggestion.

File: src/use/use.py (hunk 5 above)
Comment (greyblue9) [reply src/use/use.py]: Should be fixed

File: src/use/use.py (hunk 1 above)
Comment (greyblue9) [reply src/use/use.py]: Should be fixed

File: src/use/use.py (hunk 4 above)
Comment (greyblue9) [reply src/use/use.py:183]: Actually I think thus was one of the first few warnings when I turned on type checking, that's why I got to it :)

File: src/use/use.py (hunk 3 above)
Comment (greyblue9) [reply src/use/use.py:62]: Imported wraps() function, I can add 'import functools' back as well if you like

File: setup.py (hunk 2 above)
Comment (greyblue9) [reply setup.py:26]: Done!

**Pull request merged**